ckan/views/__init__.py

```diff
@@ -43,31 +52,202 @@ def set_cors_headers_for_response(response: Response) -> Response:
     return response
 
 
+allowed_status_codes = frozenset({HTTPStatus.OK,  # 200
+                                  HTTPStatus.MOVED_PERMANENTLY,  # 301
+                                  HTTPStatus.FOUND,  # 302
+                                  HTTPStatus.UNAUTHORIZED,  # 401
+                                  HTTPStatus.FORBIDDEN,  # 403
+                                  HTTPStatus.NOT_FOUND  # 404
+                                  })
+
+
+# ETag's are very useful since it allows conditional freshness checks
+# without bandwidth costs by the browser/cdn.
+# Since html pages from ckan are always generated on the fly it
+# can't use last modified and content length our Nginx does it
+# This is a simple etag which changes on each request
+def set_etag_for_response(response: Response) -> Response:
+    """Set ETag and return 304 if content is unchanged."""
+
+    # skip streaming content
+    if response.is_streamed:
+        return response
+
+    enable_etags = config.get(u'ckan.etags.enabled', True)
+    if enable_etags and response.status_code in allowed_status_codes:
+        if 'etag' not in response.headers:
+            etag_replace = getattr(g, 'etag_replace', None)
+            if etag_replace:
+                response.set_etag(etag_replace)
+            else:
+                # s3 etag uses md5 if you want that, load `etag` plugin, this is fast
+                # hash etag which h.set_etag_modified_time(str) can be used to set
+                # correct modified time as well as extended by plugins via
+                # `h.etag_append(str)` for their uniqueness constraints
+                mtime = getattr(g, 'etag_modified_time', time.time())
```

Comment:
We need to set the modified time even if we've already provided unique data in etag_append or etag_replace?

Reply:
etag_append is for extensions to add their own uniquness hash.
etag_replace just blats everything with whatever was set (nuclear approach) only for extensions to use (with caution)

etag_modified_time can be used in the future when extensions are supporting the etag_append feature to allow us to just use the date in the 'db' for when the record was last modified on certain pages/functions.